redfish-exporter/docs/configuration.md
# Configuration Guide

## Environment Variables

The AMD Redfish Exporter can be configured using the following environment variables:

- `UPDATED`: Last update date of the exporter
- `DESCRIPTION`: Description of the exporter
- `LISTENER_IP`: IP address to listen on (default: "127.0.0.1")
- `LISTENER_PORT`: Port to listen on for Redfish events (default: "8080")  (API events?)
- `USE_SSL`: Enable SSL/TLS (default: "false")
- `CERTFILE`: Path to SSL certificate file
- `KEYFILE`: Path to SSL key file
- `METRICS_PORT`: Port to expose Prometheus metrics (default: "2112")
- `SLURM_USER`: SLURM user name
- `SLURM_TOKEN`: SLURM authentication token
- `SLURM_CONTROL_NODE`: SLURM control node address
- `SUBSCRIPTION_PAYLOAD`: JSON string containing Redfish subscription details
- `REDFISH_SERVERS`: JSON array of Redfish server configurations
- `TRIGGER_EVENTS`: JSON array of events that trigger actions

## Redfish Server Configuration

The `REDFISH_SERVERS` environment variable should contain a JSON array of objects with the following structure:

```json
[
  {
    "ip": "https://redfish-server1.example.com",
    "username": "admin",
    "password": "password123",
    "loginType": "basic",
    "slurmNode": "compute-node-1"
  },
  {
    "ip": "https://redfish-server2.example.com",
    "username": "admin",
    "password": "password456",
    "loginType": "basic",
    "slurmNode": "compute-node-2"
  }
]
```

**NOTE:** Ensure the node name provided in `REDFISH_SERVERS` matches the SLURM node configuration. If the SLURM node name is entered incorrectly the node drain operation will fail.

## Subscription Payload

The `SUBSCRIPTION_PAYLOAD` environment variable should contain a JSON object with the following structure:

```json
{
  "Destination": "https://exporter.example.com:8080",
  "EventTypes": ["Alert", "ResourceAdded", "ResourceRemoved", "ResourceUpdated", "StatusChange"],
  "Context": "ExporterSubscription",
  "Protocol": "Redfish"
}
```

## Trigger Events

The `TRIGGER_EVENTS` environment variable should contain a JSON array of objects with the following structure:

```json
[
  {
    "Severity": "Critical",
    "Action": "DrainNode"
  },
  {
    "Severity": "Warning",
    "Action": "LogEvent"
  }
]
```

## Formatting JSON Objects

For environment variables that require JSON objects (such as `REDFISH_SERVERS`, `SUBSCRIPTION_PAYLOAD`, and `TRIGGER_EVENTS`), you can use the following one-liner to properly fomat a JSON object for use in `.env`:

```bash
echo MY_JSON=\'$(cat "path/to/your/json_file.json" | tr -d '\n' | sed "s/'/'\\\\''/g")\'  >> .env
```

Replace `MY_JSON` with the appropriate environment variable name, and `path/to/your/json_file.json` with the path to your JSON file.

### Example

- Create a JSON file with the required configuration. For example, `redfish_servers.json`:

```json
[
 {
   "ip": "https://redfish-server1.example.com",
   "username": "admin",
   "password": "password123",
   "loginType": "basic",
   "slurmNode": "compute-node-1"
 },
 {
   "ip": "https://redfish-server2.example.com",
   "username": "admin",
   "password": "password456",
   "loginType": "basic",
   "slurmNode": "compute-node-2"
 }
]
```

- Use the following one-liner to add the properly formatted environment variable to the `.env` configuration file:

```bash
echo REDFISH_SERVERS=\'$(cat "redfish_servers.json" | tr -d '\n' | sed "s/'/'\\\\''/g")\' >> .env
```

This command does the following:

- `cat "redfish_servers.json"`: Reads the content of the JSON file
- `tr -d '\n'`: Removes all newline characters, making it a single line
- `sed "s/'/'\\\\''/g"`: Escapes any single quotes in the JSON
- `echo <command> >> .env` Appends the output to the .env file

Repeat this process for other JSON object variables like `SUBSCRIPTION_PAYLOAD` and `TRIGGER_EVENTS`.

## SSL/TLS Configuration

To enable SSL/TLS:

1. Set `USE_SSL` to "true"
2. Provide paths to your SSL certificate and key files using `CERTFILE` and `KEYFILE` respectively

## SLURM Integration

To enable SLURM integration:

1. Provide the SLURM authentication token in `SLURM_TOKEN`
2. Specify the SLURM control node address in `SLURM_CONTROL_NODE`


redfish-exporter/config.go
/**
 * Copyright 2024 Advanced Micro Devices, Inc.  All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
**/

package main

import (
	"crypto/tls"
	"encoding/json"
	"log"
	"os"
	"strconv"
	"strings"

	"github.com/joho/godotenv"
)

const (
	DefaultListenerPort   = "8080"
	DefaultMetricsPort    = "2112"
	DefaultUseSSL         = "false"
	DefaultSeverityConfig = "Fatal,Critical,Informational"
)

type Config struct {
	Information struct {
		Updated     string
		Description string
	}
	SystemInformation struct {
		ListenerIP   string
		ListenerPort string
		UseSSL       bool
		MetricsPort  int
	}
	CertificateDetails struct {
		CertFile string
		KeyFile  string
	}
	SlurmToken          string
	SlurmControlNode    string
	SlurmUser           string
	SubscriptionPayload SubscriptionPayload
	RedfishServers      []RedfishServer
	TriggerEvents       []TriggerEvent
	PrometheusConfig    PrometheusConfig
	context             *tls.Config
	eventCount          int
	dataBuffer          []byte
}

type TriggerEvent struct {
	Severity string `json:"Severity"`
	Action   string `json:"Action"`
}

type PrometheusConfig struct {
	Severity []string `json:"Severity"`
}

func setupConfig() Config {
	// Load .env file
	err := godotenv.Load()
	if err != nil {
		log.Println("Failed to load .env file. Assuming environment variables are set.")
	}

	// Initialize Config object
	var AppConfig Config

	// Read environment variables
	AppConfig.Information.Updated = os.Getenv("UPDATED")
	AppConfig.Information.Description = os.Getenv("DESCRIPTION")
	AppConfig.SystemInformation.ListenerIP = os.Getenv("LISTENER_IP")

	// Read and parse LISTENER_PORT with a default value
	listenerPort := os.Getenv("LISTENER_PORT")
	if listenerPort == "" {
		listenerPort = DefaultListenerPort
	}
	AppConfig.SystemInformation.ListenerPort = listenerPort

	// Metrics Port Configuration
	metricsPortStr := os.Getenv("METRICS_PORT")
	if metricsPortStr == "" {
		metricsPortStr = DefaultMetricsPort
	}
	metricsPort, err := strconv.Atoi(metricsPortStr)
	if err != nil {
		log.Fatalf("Failed to parse METRICS_PORT: %v", err)
	}
	AppConfig.SystemInformation.MetricsPort = metricsPort
	// Read and parse USE_SSL with a default value
	useSSLStr := os.Getenv("USE_SSL")
	if useSSLStr == "" {
		useSSLStr = DefaultUseSSL
	}
	useSSL, err := strconv.ParseBool(useSSLStr)
	if err != nil {
		log.Fatalf("Failed to parse USE_SSL: %v", err)
	}
	AppConfig.SystemInformation.UseSSL = useSSL

	AppConfig.CertificateDetails.CertFile = os.Getenv("CERTFILE")
	AppConfig.CertificateDetails.KeyFile = os.Getenv("KEYFILE")

	AppConfig.SlurmToken = os.Getenv("SLURM_TOKEN")
	AppConfig.SlurmControlNode = os.Getenv("SLURM_CONTROL_NODE")
	AppConfig.SlurmUser = os.Getenv("SLURM_USER")

	subscriptionPayloadJSON := os.Getenv("SUBSCRIPTION_PAYLOAD")
	if err := json.Unmarshal([]byte(subscriptionPayloadJSON), &AppConfig.SubscriptionPayload); err != nil {
		log.Fatalf("Failed to parse SUBSCRIPTION_PAYLOAD: %v", err)
	}

	triggerEventsJSON := os.Getenv("TRIGGER_EVENTS")
	if triggerEventsJSON != "" {
		err = json.Unmarshal([]byte(triggerEventsJSON), &AppConfig.TriggerEvents)
		if err != nil {
			log.Fatalf("Failed to unmarshal TRIGGER_EVENTS: %v", err)
		}
	}

	prometheusConfigJSON := os.Getenv("PROMETHEUS_CONFIG")
	if prometheusConfigJSON != "" {
		err = json.Unmarshal([]byte(prometheusConfigJSON), &AppConfig.PrometheusConfig)
		if err != nil {
			log.Fatalf("Failed to unmarshal PROMETHEUS_CONFIG: %v", err)
		}
	}
	if prometheusConfigJSON == "" {
		AppConfig.PrometheusConfig.Severity = strings.Split(DefaultSeverityConfig, ",")
	}

	// Read and parse the REDFISH_SERVERS environment variable
	redfishServersJSON := os.Getenv("REDFISH_SERVERS")
	if redfishServersJSON == "" {
		log.Println("REDFISH_SERVERS environment variable is not set or is empty")
		return AppConfig
	}
	if err := json.Unmarshal([]byte(redfishServersJSON), &AppConfig.RedfishServers); err != nil {
		log.Fatalf("Failed to parse REDFISH_SERVERS: %v", err)
	}

	return AppConfig
}
